### PR TITLE
Add LAUNCH_EXT_NETWORKS_IPV4

### PR DESCRIPTION
**Pull request by beornf**
This PR adds two useful options as static ip addresses are not supported in swarm mode: https://github.com/moby/moby/issues/24170.

### CODE REVIEW

**Comment by tlex** [comment]
First of all, thank you for your merge request.

Looking at the [documentation](https://docs.docker.com/compose/compose-file/compose-file-v3/#ipv4_address-ipv6_address), I see a series of problems with this:
* `ipv6_address` can't work anyhow, since on [line 116](https://github.com/ix-ai/swarm-launcher/blob/master/entrypoint.sh#L116) the version defined is `3.8` (and `ipv6_address` requires v2)
* `ipv4_address` requires that the `networks` section [on line 261](https://github.com/ix-ai/swarm-launcher/blob/master/entrypoint.sh#L261) has an `ipam` block
* for compose file v3.3 and higher, `external` can't be combined with other network configuration keys (like the `ipam` from above) ([source](https://docs.docker.com/compose/compose-file/compose-file-v3/#external-1))

The more I read the docs, the more I realize that this would require a major rework (probably among the lines of deciding between v2 and v3 and then ensuring that the remaining options are actually compatible with the selected version).

**Comment by beornf** [comment]
It's true the `ipam` section is required to define the subnet when creating new networks. This PR only addresses `external` networks which works without `ipam` as the subnet can be defined externally.

I can drop the `ipv6_address` as I only added that for consistency and didn't realize the lack of support in compose v3.

If you want I can add support for static ips on the internal bridge networks by adding an `ipv4_subnet` variable. The logic being for internal networks `ipv4_address` and `ipv4_subnet` is required and external networks only `ipv4_address` is needed.

Maintaining and testing multiple versions of compose sounds like adding too much complexity to me?

**Comment by tlex** [comment]
Yeah, would be a nightmare to keep tracking what gets implemented when.

I can accept `ipv4_address`, however there's one case that needs to be covered: combining `LAUNCH_NETWORKS` (see [L225](https://github.com/ix-ai/swarm-launcher/blob/master/entrypoint.sh#L255)) and `LAUNCH_EXT_NETWORKS` (see [L258](https://github.com/ix-ai/swarm-launcher/blob/master/entrypoint.sh#L258)) needs to be handled, since you can't mix a list with a dict. So I suggest changing [L225](https://github.com/ix-ai/swarm-launcher/blob/master/entrypoint.sh#L255) similarly to [L263](https://github.com/ix-ai/swarm-launcher/pull/26/files#diff-6f9d41d046756f0ddc2fcee0626bdb50100d12b88f293734eff742818e03efa2R261) in your PR.

File: README.md [hunk 1]
```diff
@@ -54,6 +54,8 @@ The following environment variables are important if you don't supply a `/docker
 | `LAUNCH_SERVICE_NAME`   | random (by swarm-launcher) | NO            | If you want to use a specific name for the service |
 | `LAUNCH_CONTAINER_NAME` | random (by docker)         | NO            | If you want to use a specific name for the container (similar to the task name) |
 | `LAUNCH_HOSTNAME`       | -                          | NO            | If you want to use a specific hostname for the container |
+| `LAUNCH_IPV4_ADDRESS`   | -                          | NO            | If you want to use a specific IPv4 address for the container |
+| `LAUNCH_IPV6_ADDRESS`   | -                          | NO            | If you want to use a specific IPv6 address for the container |
```
Comment (tlex) [inline README.md:58]:
```suggestion
```

File: entrypoint.sh [hunk 2]
```diff
@@ -66,6 +66,8 @@ if [ -f ${COMPOSE_FILE} ]; then
     'LAUNCH_CONTAINER_NAME'
     'LAUNCH_PRIVILEGED'
     'LAUNCH_HOSTNAME'
+    'LAUNCH_IPV4_ADDRESS'
+    'LAUNCH_IPV6_ADDRESS'
```
Comment (tlex) [inline entrypoint.sh:70]:
```suggestion
```

File: entrypoint.sh [hunk 3]
```diff
@@ -256,7 +258,13 @@ xEOF
...
       done
       for NETWORK in ${LAUNCH_EXT_NETWORKS}; do
-        echo "      - ${NETWORK}" >> ${COMPOSE_FILE}
+        echo "      ${NETWORK}:" >> ${COMPOSE_FILE}
+        if [ -n "${LAUNCH_IPV4_ADDRESS}" ]; then
+          echo "        ipv4_address: ${LAUNCH_IPV4_ADDRESS}" >> ${COMPOSE_FILE}
+        fi
+        if [ -n "${LAUNCH_IPV6_ADDRESS}" ]; then
+          echo "        ipv6_address: ${LAUNCH_IPV6_ADDRESS}" >> ${COMPOSE_FILE}
+        fi
```
Comment (tlex) [inline entrypoint.sh:267]:
```suggestion
```

File: entrypoint.sh [hunk 4]
```diff
@@ -256,7 +258,13 @@ xEOF
         echo "      - ${NETWORK}" >> ${COMPOSE_FILE}
```
Comment (tlex) [inline entrypoint.sh:258]:
```suggestion
        echo "      ${NETWORK}:" >> ${COMPOSE_FILE}
```

File: README.md [hunk 5]
```diff
@@ -54,6 +54,8 @@ The following environment variables are important if you don't supply a `/docker
 | `LAUNCH_SERVICE_NAME`   | random (by swarm-launcher) | NO            | If you want to use a specific name for the service |
 | `LAUNCH_CONTAINER_NAME` | random (by docker)         | NO            | If you want to use a specific name for the container (similar to the task name) |
 | `LAUNCH_HOSTNAME`       | -                          | NO            | If you want to use a specific hostname for the container |
+| `LAUNCH_IPV4_ADDRESS`   | -                          | NO            | If you want to use a specific IPv4 address for the container |
```
Comment (tlex) [inline README.md:57]:
```suggestion
| `LAUNCH_IPV4_ADDRESS`   | -                          | NO            | If you want to use a specific IPv4 address for the container (only works if `LAUNCH_EXT_NETWORKS` is set) |
```

File: entrypoint.sh [hunk 6]
```diff
@@ -256,7 +258,13 @@ xEOF
...
       done
       for NETWORK in ${LAUNCH_EXT_NETWORKS}; do
-        echo "      - ${NETWORK}" >> ${COMPOSE_FILE}
+        echo "      ${NETWORK}:" >> ${COMPOSE_FILE}
+        if [ -n "${LAUNCH_IPV4_ADDRESS}" ]; then
+          echo "        ipv4_address: ${LAUNCH_IPV4_ADDRESS}" >> ${COMPOSE_FILE}
```
Comment (tlex) [inline entrypoint.sh:263]: Now I realize - the same IP address will be set for *all* the `LAUNCH_EXT_NETWORKS`. This must be handled differently.

**Review by tlex — CHANGES_REQUESTED** [review]
I've finished now the review, however I worry about setting the same IP address on multiple `LAUNCH_EXT_NETWORKS`. I see a few possibilities here: either the IP address is set only for the first network in the list (and this needs to be documented in the README), or we find a way to include as a separate variable (e.g. in the form `LAUNCH_EXT_NETWORKS_IPV4=mynetwork:10.0.0.1`)

**Comment by beornf** [comment]
I've used a couple of bashisms to split the separators in `LAUNCH_EXT_NETWORKS_IPV4`. Can change them to unix tools like `cut`, `sed` or `awk` if you have a preference. I could rewrite the inner loop with associative arrays but most of the time it would only be a few elements. 🤔

**Comment by tlex** [comment]
Thanks for your work.

I've built a new version of `registry.gitlab.com/ix.ai/swarm-launcher:dev-branch` with the changes included. I've also added one more commit (bb71cee) that takes a similar approach to converting `Network:IP Network2:IP2` as it was implemented with the `LAUNCH_ULIMITS`.

You can also take it for a spin and see if it works.

**Comment by beornf** [comment]
The MR looks much better than mine especially making the `LAUNCH_EXT_NETWORKS` and `LAUNCH_EXT_NETWORKS_IPV4` variables independent. Checked out the Gitlab code and confirmed its working for my swarm cluster. 👍

**Comment by tlex** [comment]
Thanks for the feedback. I've release `v0.14.0` with the new feature.

**Pull request merged**